polkadot/xcm/xcm-builder/src/asset_exchange.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Adapters that implement the [`xcm_executor::traits::AssetExchange`] trait.
+
+use core::marker::PhantomData;
+use frame_support::traits::tokens::fungibles;
+use pallet_asset_conversion::{QuoteExchangePrice, SwapCredit};
+use sp_runtime::traits::Zero;
+use sp_std::vec;
+use xcm::prelude::*;
+use xcm_executor::{
+	traits::{AssetExchange, MatchesFungibles},
+	AssetsInHolding,
+};
+
+/// An adapter from [`pallet_asset_conversion::SwapCredit`] and
+/// [`pallet_asset_conversion::QuoteExchangePrice`] to [`xcm_executor::traits::AssetExchange`].
+///
+/// Takes just one fungible asset in `give` and allows only one fungible asset in `want`.
+pub struct FungiblesPoolAdapter<AssetConversion, Fungibles, Matcher, AccountId>(
+	PhantomData<(AssetConversion, Fungibles, Matcher, AccountId)>,
+);
+impl<AssetConversion, Fungibles, Matcher, AccountId> AssetExchange
+	for FungiblesPoolAdapter<AssetConversion, Fungibles, Matcher, AccountId>
+where
+	AssetConversion: SwapCredit<
+			AccountId,
+			Balance = u128,
+			AssetKind = Fungibles::AssetId,
+			Credit = fungibles::Credit<AccountId, Fungibles>,
+		> + QuoteExchangePrice<Balance = u128, AssetKind = Fungibles::AssetId>,
+	Fungibles: fungibles::Balanced<AccountId, Balance = u128>,
+	Matcher: MatchesFungibles<Fungibles::AssetId, Fungibles::Balance>,
+{
+	fn exchange_asset(
+		_: Option<&Location>,
+		give: AssetsInHolding,
+		want: &Assets,
+		_: bool,
```

Comment:
This implementation basically with `maximal = false`. It swaps exactly as set in `want`. Also this implementation care exactly about first asset in `give` and `want`.

Reply:
I think for a correct implementation it should at least error when not all `want` are supplied:
```rust
	/// - `want`: The minimum amount of assets which should be given to the caller in case any
	///   exchange happens. If more assets are provided, then they should generally be of the same
	///   asset class if at all possible.
```

and also if `maximal` is true then at least try to convert every fungible from `give` to all of `want`:
```rust
	/// - `maximal`: If `true`, then as much as possible should be exchanged.
```

Reply:
Added the `maximal = true` case but yes, this implementation uses only the first asset, I'll change the name to reflect that.

Reply:
@acatangiu Any names come to mind?

Reply:
Yes it looks good to me, but I still think it should return an error if there is more than one asset in `want`